build.py

```diff
@@ -34,36 +62,43 @@ def print_usage():
             name += " (default with --debug)"
         configs.append(name + "\n    " + " ".join(args))
     configs_string = "\n  ".join(configs)
-    cmake_name = os.path.basename(CMAKE)
-    generator_name = CMAKE_GENERATOR.lower()
-    default_config_name = DEFAULT_CONFIG_NAME
-    debug_config_name = DEBUG_CONFIG_NAME
-    print(f"""Usage: {script_name} [BUILD [BUILD ...]] [--all] [--debug] [MAKE_OPTIONS]
-
-Build one or more predefined build configurations of Fast Downward. Each build
-uses {cmake_name} to compile the code using {generator_name} . Build configurations
-differ in the parameters they pass to {cmake_name}. By default, the build uses all
-available cores if this number can be determined. Use the "-j" option for
-{cmake_name} to override this default behaviour.
-
-Build configurations
+    epilog = f"""build configurations:
   {configs_string}
 
---all         Alias to build all build configurations.
---debug       Alias to build the default debug build configuration.
---help        Print this message and exit.
-
-Make options
-  All other parameters are forwarded to the build step.
-
-Example usage:
-  ./{script_name}                     # build {default_config_name} in #cores threads
-  ./{script_name} -j4                 # build {default_config_name} in 4 threads
-  ./{script_name} debug               # build debug
-  ./{script_name} --debug             # build {debug_config_name}
-  ./{script_name} release debug       # build release and debug configs
-  ./{script_name} --all VERBOSE=true  # build all build configs with detailed logs
-""")
+example usage:
+  {script_name}                     # build {DEFAULT_CONFIG_NAME} with #cores parallel processes
+  {script_name} -j4                 # build {DEFAULT_CONFIG_NAME} with 4 parallel processes
+  {script_name} debug               # build debug
+  {script_name} --debug             # build {DEBUG_CONFIG_NAME}
+  {script_name} release debug       # build release and debug configs
+  {script_name} --all VERBOSE=true  # build all build configs with detailed logs
+"""
+
+    parser = argparse.ArgumentParser(
+        description=description, epilog=epilog, formatter_class=RawHelpFormatter)
+    parser.add_argument("--debug", action="store_true",
+                        help="alias to build the default debug build configuration")
+    parser.add_argument("--all", action="store_true",
+                        help="alias to build all build configurations")
+    remaining_args = parser.add_argument("arguments", nargs=argparse.REMAINDER,
+                        help="build configurations (see below) or build options")
+    remaining_args.completer = complete_arguments
+
+    if HAS_ARGCOMPLETE:
+        argcomplete.autocomplete(parser)
+    args, unparsed_args = parser.parse_known_args()
+
+    if hasattr(args, "remaining_args"):
```

Comment:
Can this case ever arise? How?

Reply:
This happens with the command that Salome posted above: `build.py release -j4`. Our fix was to use `parse_known_args` instead which means that `remaining_args` remains non-existent and `-j4` ends up in `unparsed_args`. I'll add a comment

Reply:
Ahh, I see you meant the other case. You're right, this should not happen because everything except the known options (`--all`, `--debug`) would end up in `unparsed_args`. In that case, having the option with `nargs=argparse.REMAINDER` is redundant. We could remove it but it would affect the generated help text: the description of the positional arguments was bad anyway and we could improve it this way, but the "using" statement would lose the meta vars `[BUILD ...] [CMAKE_OPTION ...]` in the end. Is that OK?

Reply:
I guess we have to keep the argument because this is where our argument completer attaches to. Removing it would not be an option.

Reply:
I keep finding problems with this code: `remaining_args` is the name of the argparse representation of the argument, not the name of the parsed values. Argparse always creates the actual value, so the whole check isn't necessary, sorry for the noise. Here is the updated version:

```Python
    # With calls like 'build.py release -j4' the parser would try to interpret
    # '-j4' as an option and fail to parse. We use parse_known_args to parse
    # everything we recognize and add everything else to the list of arguments. 
    args, unparsed_args = parser.parse_known_args()
    args.arguments += unparsed_args
```

Reply:
That's the direction I meant, yes. But the example is wrong, unless I'm now really confused. `build.py release -j4` would work with regular argparse parsing. After the first positional argument, argument parsing no longer looks for options, so the example would be fine because seeing `release` means that we're now past the options and everything that follows is a positional argument. The failing cases would be those without explicit builds listed but with CMake options, for example `./build.py -j4`.

Note that if we switch to `parse_known_args`, there is a difference of behaviour to the old code: in the old code "./build.py -j4 --all" would be illegal because we expect the CMake options at the end. With `parse_known_args`, they can occur anywhere, which isn't necessarily a problem for build.py, but something to be aware of. (This may be the reason why we avoided argparse for build.py before this pull request.)

In the translator one reason we decided to go for the syntax "driver_options filenames component_options" is because the filenames always put a buffer in between the two types of options, so we can use the regular argparse behaviour of assuming the parsed options end the first time we encounter an argument not starting with "-". (Of course excluding arguments that are parameters of options, like the 400s in `--overall-time-limit 400s`.) Part of the rationale is that if someone writes a typo as in `--overall-timelimit 400s`, we don't want this to be silently converted to a search component option, which would run the translator and then abort search with a confusing error message.

I don't mind mixing the different kinds of arguments in build.py, as it has a simpler syntax than fast-downward.py, but I think the proposed code has another, more subtle bug. Normally options are parsed from left to right and can override each other, so if I call "make -j4 -j8", this means using 8 jobs (not tested, but this is the general convention because it makes it possible to do stuff like `alias make=make -j4` to set a default while still allowing this default to be overridden). With the proposed code, if I write `./build.py -j4 --debug -j8`, we would set `args.arguments` to `["-j8", "-j4"]`, inverting the order. The values we get from `unparsed_args` will always appear in the command line before the values we get from the positional `args.arguments`, so I would replace the `+=` with something like `args.arguments = unparsed_args + args.arguments` if we want to recycle `args.arguments` in the first place and not assign the concatenation to a fresh variable.

(This could be done in-place as `args.arguments[0:0] = unparsed_args`, but this looks too clever.)

Reply:
Actually, I'm not 100% sure if it's still the case that the first positional arguments finishes the search for options if `parse_known_args` is used. So not sure if it is indeed the case that we can know in which order the arguments occurred. (I think I know how the "usual" arg parsing mode works, but I'm a bit less sure about `parse_known_args`.)

Reply:
OK, I tested it: indeed also with known arg parsing, once something looks like a positional argument, the rest will show up in the remainder.

Test code:
```
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--debug", action="store_true")
parser.add_argument("--all", action="store_true")
parser.add_argument("arguments", nargs=argparse.REMAINDER)
args, unparsed_args = parser.parse_known_args()

print(f"args.arguments: {args.arguments}")
print(f"unparsed_args: {unparsed_args}")
```

Example run:
```
$ ./foo.py -j4 --debug foo --all bar
args.arguments: ['foo', '--all', 'bar']
unparsed_args: ['-j4']
```
So `foo` causes the rest to appear in the remainder `args.arguments`, even though a legal option (`--all`) appears later. This means that indeed all arguments in `unparsed_args` must come before all arguments in `args.arguments`.

Reply:
Indeed `./build.py -j4 --all` was legal before argparse, so I think it is good that it is legal again with using `parse_known_args`. I went with `args.arguments = unparsed_args + args.arguments`